### PR TITLE
Fix/sdl crash on shutdown issues

### PR DESCRIPTION
**Pull request by OHerasym**
### PR Changes

- Remove unused code from utils
- Fix logger shutdown
- Fix catching signal SIGTERM
- Add class ThreadCounter for debugging multi thread issues

Related: [APPLINK-25734](https://adc.luxoft.com/jira/browse/APPLINK-25734) , [APPLINK-26736](https://adc.luxoft.com/jira/browse/APPLINK-26736)

### CODE REVIEW

**Comment by OHerasym** [comment]
Please review: @orhan-mehmedov , @Kozoriz , @anosach-luxoft , @LuxoftAKutsan , @VVeremjova

**Comment by Kozoriz** [comment]
Reviewed

File: src/components/utils/include/utils/threads/thread_counter.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ */
+
+#ifndef SRC_COMPONENTS_UTILS_INCLUDE_UTILS_THREAD_COUNTER_H_
+#define SRC_COMPONENTS_UTILS_INCLUDE_UTILS_THREAD_COUNTER_H_
+
+#include "utils/threads/thread.h"
+
+namespace threads {
+
+class ThreadCounter {
```
Comment (orhan-mehmedov) [inline src/components/utils/include/utils/threads/thread_counter.h]: @OHerasym, It'll be nice to have some description here (doxygen) about that class.

File: src/components/utils/src/threads/thread_posix.cc [hunk 2]
```diff
@@ -275,10 +278,13 @@ void Thread::join() {
 }
 
 Thread::~Thread() {
-  finalized_ = true;
-  stopped_ = true;
-  join();
-  pthread_join(handle_, NULL);
+  if (!finalized_) {
```
Comment (orhan-mehmedov) [inline src/components/utils/src/threads/thread_posix.cc]: @OHerasym, I do not see value of finalized_ to be changed elsewhere but here, so this if() will always be true. Probably there is some future/design extension/intention? If someone have info - please share.

File: src/components/utils/src/threads/thread_posix.cc [hunk 3]
```diff
@@ -275,10 +278,13 @@ void Thread::join() {
...
+    ThreadCounter::Decrement();
+    finalized_ = true;
+    stopped_ = true;
```
Comment (orhan-mehmedov) [inline src/components/utils/src/threads/thread_posix.cc]: @OHerasym, stopped_ = true; will be set again to true in join() -> stop() -> stopeed_ = true. What is the intention to set it true here?

File: src/components/utils/src/threads/thread_posix.cc [hunk 4]
```diff
@@ -275,10 +278,13 @@ void Thread::join() {
...
+    join();
+    pthread_join(handle_, NULL);
```
Comment (orhan-mehmedov) [inline src/components/utils/src/threads/thread_posix.cc]: @OHerasym, Isn't it better to check here thread_options_.is_joinable() before calling pthread_join(handle_, NULL);?

File: src/components/utils/src/threads/thread_posix.cc (hunk 2 above)
Comment (OHerasym) [inline src/components/utils/src/threads/thread_posix.cc]: @orhan-mehmedov , finalized_ will be changed inside if, it garanties that pthread_join() will be called only ones.

File: src/components/utils/src/threads/thread_posix.cc (hunk 3 above)
Comment (OHerasym) [inline src/components/utils/src/threads/thread_posix.cc]: @orhan-mehmedov, yes I saw this, i just added if case in destructor, didn't remove this, probably there is some case, were it used. Didn't checked.

File: src/components/utils/src/threads/thread_posix.cc (hunk 4 above)
Comment (OHerasym) [inline src/components/utils/src/threads/thread_posix.cc]: @ orhan-mehmedov , 98bb36d , Done

File: src/components/utils/include/utils/threads/thread_counter.h (hunk 1 above)
Comment (OHerasym) [inline src/components/utils/include/utils/threads/thread_counter.h]: @orhan-mehmedov , 98bb36d , Done

**Comment by orhan-mehmedov** [comment]
@OHerasym, Reviewed.

**Comment by VVeremjova** [comment]
Reviewed

**Comment by pestOO** [comment]
@OHerasym,
please don;t use temp commit messages.
In truly recommend to use [`--fixup` commits](https://robots.thoughtbot.com/autosquashing-git-commits)

File: src/components/utils/src/logger_posix.cc [hunk 5]
```diff
@@ -108,6 +108,16 @@ void logger::Logger::Impl::DeinitLogger() {
   set_log_level(LogLevel::LL_TRACE);
   delete message_loop_thread_;
   message_loop_thread_ = NULL;
+  log4cxx::LoggerPtr rootLogger = log4cxx::Logger::getRootLogger();
+  log4cxx::spi::LoggerRepositoryPtr repository =
+      rootLogger->getLoggerRepository();
+  log4cxx::LoggerList loggers = repository->getCurrentLoggers();
+  for (log4cxx::LoggerList::iterator i = loggers.begin(); i != loggers.end();
+       ++i) {
+    log4cxx::LoggerPtr logger = *i;
```
Comment (AGaliuzov) [inline src/components/utils/src/logger_posix.cc]: @OHerasym Are you sure you need to create this pointer over and over again?

File: src/components/utils/src/signals_posix.cc [hunk 6]
```diff
@@ -50,10 +50,13 @@ void SigHandler(int sig) {
       SDL_INFO("SIGTERM signal has been caught");
       break;
     case SIGSEGV:
+      SDL_FLUSH_LOGGER();
       SDL_INFO("SIGSEGV signal has been caught");
+      abort();
```
Comment (AGaliuzov) [inline src/components/utils/src/signals_posix.cc]: @OHerasym why do we need this abort?

File: src/components/utils/src/threads/thread_counter.cc [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+#include "utils/threads/thread_counter.h"
+#include "utils/logger.h"
+
+namespace threads {
+
+SDL_CREATE_LOGGER("Utils")
+
+ThreadCounter::ThreadCounter() {
+  DCHECK(ThreadCounter::GetValue() == 0);
```
Comment (AGaliuzov) [inline src/components/utils/src/threads/thread_counter.cc]: @OHerasym what the expected behaviour in `release` build? The same question for destructor

File: src/components/utils/src/threads/thread_win.cc [hunk 8]
```diff
@@ -121,7 +121,11 @@ Thread::Thread(const char* name, ThreadDelegate* delegate)
     , isThreadRunning_(false)
     , stopped_(false)
     , finalized_(false)
-    , thread_created_(false) {}
+    , thread_created_(false) {
+#ifdef THREAD_COUNT
+  ThreadCounter::Increment();
```
Comment (AGaliuzov) [inline src/components/utils/src/threads/thread_win.cc]: @OHerasym do we have the common interface for all our threads? It seems like we can move this code to base's class ctor/dtor

File: src/components/utils/src/logger_posix.cc [hunk 9]
```diff
@@ -108,6 +108,16 @@ void logger::Logger::Impl::DeinitLogger() {
   set_log_level(LogLevel::LL_TRACE);
   delete message_loop_thread_;
   message_loop_thread_ = NULL;
+  log4cxx::LoggerPtr rootLogger = log4cxx::Logger::getRootLogger();
```
Comment (LuxoftAKutsan) [inline src/components/utils/src/logger_posix.cc]: @OHerasym I think this changes required description

**Comment by OHerasym** [comment]
Pull request will be reopened after vacation

**Pull request closed without merging**